Update the importing spec

stefnotch opened the pull request:
This updates the importing spec

ncthbrt commented:
Sorry. Didn't mean to mark as ready

stefnotch commented:
(No worries, I'll just update the title)

k2d222 commented:
just a couple comments

the `crate::` keyword may not be the best choice, as crates are inherently a rust concept. I suggest `package` or `root` (preferred).

in rust 2018 edition, an import which starts with an ident (not `crate`, `self` or `super`) first looks at the child modules, then at external crates. Should we do that as well? The downside is that it costs a filesystem lookup and is ambiguous. The upside is that it simplifies `import self::foo::bar` to `import foo::bar`.

mighdoll commented:
> The `crate::` keyword may not be the best choice, as crates are inherently a rust concept. I suggest `package` or `root`.

I tried to summarize the then-current discussion in this [hack.md doc](https://hackmd.io/ljkByEcnQa2NdNLWed2M6Q#root-shortcut-for-import-statements
). There we used `package` and `super` (but not `self` or `crate`).

> first looks at the child modules, then at external crates. Should we do that as well? The downside is that it costs a filesystem lookup and is ambiguous. 

Simplifying a common import case seems goodness, though it's also tempting to defer until after M1 if it's nontrivial. I was thinking that most of the features in this [section](https://hackmd.io/ljkByEcnQa2NdNLWed2M6Q#Possible-Future-Features-in-the-Module-System) could be deferred til after our first rev.

stefnotch commented:
I updated the importing spec. @k2d222 I also changed the grammar a tiny bit, after realizing that I made a mistake and unintentionally disallowed `import super::foo`.

stefnotch commented on Imports.md:
```diff
@@ -20,268 +14,230 @@ Finally, we want **multiple tools** which can compile WGSL-with-imports down to
-import_path:  
-| ident '/' (import_path | import_collection | item_import | star_import)  
+import_relative:
+| ('package' | 'super') '::' ('super' '::')*
```
Yes, I didn't find a strong enough reason to keep for M1.

stefnotch commented on Imports.md:
```diff
@@ -20,268 +14,230 @@ Finally, we want **multiple tools** which can compile WGSL-with-imports down to
-Dead code elimination is allowed, but not required.
+## Dead Code Elimination
+
+Linkers may choose to do dead code elimination, but it is a not-observable implementation detail.
```
Thanks for pointing this out! I added a note regarding them.

stefnotch commented on Imports.md:
```diff
@@ -20,268 +14,230 @@ Finally, we want **multiple tools** which can compile WGSL-with-imports down to
+## Entry points and pipeline overridable constants
 
-These items are preserved when importing a module. They are not imported, but will land in the final module with a mangled name.
+These items are preserved when importing a module. Their name must be preserved. They will land in the final module, if they are being referenced.
```
I changed it to use that wording 👍

stefnotch commented:
> one thing missing I think is describing where to look for the files:
> 
>     1. there is a root folder provided to the linker.

Should that be a root folder, or a root file? As in, should it be possible to write `package::Foo` to refer to a `Foo` struct that is defined in the WESL equivalent of a `lib.rs`?

>     2. how a relative path with self or super is converted to absolute path that starts at the root

I added this. Let me know if that's clear enough
"To get an absolute path to a module, one follows the algorithm above. In step 1, one takes the known absolute path of the `super` module, or the package.
The absolute path of the `super` module is always known, since the first loaded WESL file must always be the root module, and children are only discovered from there."



>     3. what happens if you have too many supers and exit the root?

Exiting the root should be an error


>     4. can we load both .wesl and .wgsl files? which takes priority?

I added a cop-out section at the end, which says "only wesl for now"

mighdoll commented on Imports.md:
```diff
@@ -417,3 +362,7 @@ See [Conditional Compilation](./ConditionalCompilation.md).
+
+## Importing `.wgsl` files
+
+Currently only `.wesl` files are importable. This restriction may be relaxed in the future.
```
Why not allow importing existing .wgsl files? Importing .wgsl makes it clear that WESL tooling is easy to adopt. 

If the only issue is duplicate named files like foo.wesl and foo.wgsl, we can pick a rule (perhaps drop the foo.wgsl, and linker warn)

stefnotch replied:
I didn't want to add an additional "check for existence of different files" check to the linker rules. Node.js has that (js, cjs, mjs, ...), and from what I can gather, it was a mistake.

I would be envisioning some syntax level way of distinguishing between `import foo::bar` and `import foo::wgsl::bar`

stefnotch replied:
But you're right, we should add a way of importing wgsl files as soon as possible.

k2d222 replied:
my implementation looks for the wesl file and if it does not exist it looks for a wgsl file and treats it as a wesl file. Idk if in the future we might want to treat vanilla wgsl files differently, e.g. disallow wesl extensions in them.

stefnotch replied:
TODO: Add a compatibility mode where wgsl files are allowed.
For publishing, they can simply be converted to wesl.

stefnotch replied:
Adding this

mighdoll commented on Imports.md:
```diff
@@ -370,44 +333,26 @@ For example, a linker can identify imports
-We encourage tooling authors to also implement source maps when implementing imports.
+We encourage tooling authors to also implement source maps when implementing imports. This aids
+- Error Reporting. When Naga or Tint report an error in the generated WGSL code, we want to map the error location back to the WESL code.
+- Debugging. Eventually we hope to have a full toolchain of WESL ==generates a sourcemap==> WGSL ==generates a source map==> SPIR-V ==> RenderDoc shows WESL code
```
(minor) the `sourcemap==>` syntax looks a little cramped. Maybe try spaces and -> or some alternative?

stefnotch replied:
I rephrased it as normal text

mighdoll commented on Imports.md:
```diff
@@ -370,44 +333,26 @@ For example, a linker can identify imports
 
-We are slowly working on adopting and adjusting the specification.
-This work is happening in multiple tools, including [a WGSL language server](https://github.com/dannymcgee/vscode-wgsl/), a [Typescript WGSL linker](https://github.com/mighdoll/wgsl-linker) and [Bevy's WGSL-imports library](https://github.com/bevyengine/naga_oil).
+This will be implemented in the [Javascript](https://github.com/wgsl-tooling-wg/wesl-js) and [Rust](https://github.com/wgsl-tooling-wg/wesl-rs) linkers.
```
JavaScript. (cap s)

stefnotch replied:
Thanks 👍

k2d222 reviewed (approved):
Thanks for your work @stefnotch ! I'm happy with the current state, I would approve merging after adressing @mighdoll 's comments.

stefnotch reviewed (commented):
Make it clear what is in scope when I do `import foo::bar`, or when I do `import foo::{bar, baz as cat, foo}`.

Note that it's not necessary to parse dependent files when adding all those things to the module scope.

Imported items are in the module scope. They are not re-exported though.

stefnotch commented:
A glob syntax in the `wesl.toml` could make sense for
- wgsl files are allowed/ignored
- this folder has tests, don't bother
- files can be scattered all over the project

k2d222 commented:
Just thought of something. Currently the import path segments are WGSL identifiers. Maybe we want to restrict them more if they are supposed to match file/folder names. We want our packages to work on all platforms regardless of filesystem name restrictions.

This is two-fold: file names should be valid WGSL identifiers, and WESL module identifiers should be valid file names.

So for example the following should not be allowed (or just not reachable): 
* file `foo.bar.wesl` because `.` is not a valid identifier character. (same for all punctuation & symbols)
* file `foo bar.wesl` because it is two words.
* file `__foo.wesl` because WGSL idents starting with `_ _` are reserved.
* file `1foo.wesl` because WGSL idents cannot start with numbers.
* file `struct.wesl` because `struct` is a reserved WGSL word.
* perhaps folders `CON, PRN, AUX, NUL, COM1 - COM9, LPT1 - LPT9` because they are Windows reserved names?

Should we restrict to latin alphabet? Are there popular filesystems that don't support unicode letters? How do other languages deal with that issue?

k2d222 commented:
> From what I can tell, other programming languages deal with this by ignoring the problem.

If that is true, then maybe we should do exactly the same :) Or just write a short note about it.

stefnotch commented:
I added a new commit which addresses some review comments.

- Grammar change: Just swaps two elements. I'm thinking about including a version of the grammar that can be parsed with a single token lookahead, instead of two.
- Adds a note about where the root folder/root module is. (linker specific for now)
- Allows for `.wgsl` files
- Warns people of the terrors of filesystems

stefnotch commented:
Okay, I added a LL(1) grammar. @k2d222 or @mighdoll , could you double check if the grammar is the same?

Also, should I get rid of the non-LL(1) grammar? One downside of it is that it's more recursive, instead of using a `*` for repetition.

k2d222 reviewed (commented):
I vote for keeping only the LL1 to reduce noise.
Noticed that the grammar does not seem to allow `import foo;`, where foo is the root module of package foo.

stefnotch commented on Imports.md:
```diff
@@ -20,70 +14,70 @@ Finally, we want **multiple tools** which can compile WGSL-with-imports down to
-import_path:  
-| ident '/' (import_path | import_collection | item_import | star_import)  
+import_relative:
+| ('package' | 'super') '::' ('super' '::')*
```
That's very smart, thank you!

stefnotch commented:
> Noticed that the grammar does not seem to allow `import foo;`, where foo is the root module of package foo.

Should it be allowed, and what should the semantics be? That would definitely be a module import, so inline it could only be used as `foo::bar`.
But then it's equivalent to `foo::bar` without an import statement?

k2d222 commented:
@mighdoll my linker currently supports it so no additional effort on my side. I'm not going to argue if you think it's worth having  :+1: 

It might be useful for future `@publish` / re-exports too.

stefnotch commented:
Okay, I'm convinced, I'll tweak the grammar to allow that. Seems like it could be very useful in the future 👍

stefnotch commented:
Also, I'm allowing `import { foo, bar }`, because it simplifies the grammar, and Rust also allows it.

stefnotch commented on Imports.md:
````diff
@@ -20,268 +14,243 @@ Finally, we want **multiple tools** which can compile WGSL-with-imports down to
 }
 ```
 
 Both `bevy_ui` and `my` are packages in the current project. Language servers and related tools can look in a `wgsl.toml` file to find the location of the packages. This lets libraries be published to package managers, and users can import them with a simple syntax.
````
Ah, good catch! Fixed :)

mighdoll commented on Imports.md:
```diff
@@ -20,268 +14,243 @@ Finally, we want **multiple tools** which can compile WGSL-with-imports down to
-item_import:
-| ident ('as' ident)?
+import_path_or_item:
+| ident ('::' (import_collection | import_path_or_item) | 'as' ident)?
```
Perhaps something like this would be easier for the reader to parse? (I got a little crossed in the first read)
```ebnf
import_path_or_item:
| ident ('as' ident)?
| ident ('::' (import_collection | import_path_or_item))?
```

stefnotch replied:
Hmm, yes, that's easier to read. But it's no longer LL(1).
Then again, manually doing the conversion when implementing it is trivial.

I don't know what the preferred option is. @k2d222 any opinions?

stefnotch replied:
Okay, the official recursive descent grammar also has enough cases where they opted for simplicity over LL(1)

mighdoll commented on Imports.md:
```diff
@@ -20,268 +14,243 @@ Finally, we want **multiple tools** which can compile WGSL-with-imports down to
-## Examples
+To resolve the import, the recursive structure is flattened out. Then, one iterates over each segment, and looks it up one by one.
+
+1. We start with the first segment.
```
describe what flattened out means? I understand, but not sure the next reader would. Perhaps something about rewriting a path with collections to get rid of those and maybe ignoring the 'as' parts.

also note the iteration is left to right, e.g. iterations starts with the leftmost segment in the flattened import path.

stefnotch replied:
I expanded on this section a little bit. I now realize that I haven't really addressed how `as` aliases work.
I'm not entirely sure how to best describe them.

mighdoll commented on Imports.md:
```diff
@@ -20,268 +14,243 @@ Finally, we want **multiple tools** which can compile WGSL-with-imports down to
-<tr>
-<td>
+Linkers are allowed to fall back to `.wgsl` files when a `.wesl` file cannot be found *in user code*.
+Published libraries will only contain `.wesl` files. Thus, linkers should rename `.wgsl` files to `.wesl` for publishing.
```
why does the suffix need to be translated? once we've ingested the file the linker looks it up by module path and ignores the suffix anyway.. 

If it doesn't matter, it's nice to expose the interop with wgsl, and have error messages that refer to the correct source file not a suffix translated one..

mighdoll commented on Imports.md:
```diff
@@ -20,268 +14,243 @@ Finally, we want **multiple tools** which can compile WGSL-with-imports down to
 
-The steps of identifier resolution are as follows:
+TODO: Do we have an issue that I can link to?
 
```
See #65

stefnotch replied:
Thank you!

mighdoll commented on Imports.md:
```diff
@@ -370,44 +340,26 @@ For example, a linker can identify imports
+- Error Reporting. When Naga or Tint report an error in the generated WGSL code, we want to map the error location back to the WESL code.
+- Debugging. Eventually we hope to have a full toolchain of WESL to WGSL to SPIR-V, with source maps at each step. In the end, it should be possible for RenderDoc to show the original WESL code.
 
 ## Preprocessor
```
conditional spec is in since this was originally written. are there any open questions wrt imports? If not, suggest editing this

stefnotch replied:
I'll remove this bit, and if any questions ever arise, we can address them in the future 👍

mighdoll commented on Imports.md:
```diff
@@ -20,268 +14,243 @@ Finally, we want **multiple tools** which can compile WGSL-with-imports down to
 
-Dead code elimination is allowed, but not required.
+Linkers may choose to do dead code elimination, but it is a not-observable implementation detail.   
+`const_assert` statements inside of functions need special treatment. Either those functions have to be kept, 
```
Perhaps just say that we're going to drop dead functions and their `const_assert`s?
(Or make this a github issue instead of trying to explain our future options here.)

stefnotch replied:
Okay, I linked to the issue

mighdoll commented on Imports.md:
`lib.wesl` for packages? so packages can publish functions at the root level, as e.g. `random_wgsl` does. (or is there some other mechanism for that?)

stefnotch replied:
The current mechanism is having a structure like
```
wesl.toml
shaders.wesl
shaders/
  foo.wesl
  bar.wesl
```

I'd say that this might just be good enough for M1, but we should revisit it. I also intentionally left that part of the spec under specified, so our linkers can try out a few different APIs
